File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/dialog/MealSelectorDialog.java
package de.karzek.diettracker.presentation.main.diary.meal.dialog;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.model.MealDisplayModel;

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
public class MealSelectorDialog extends AppCompatDialogFragment {

    @BindView(R.id.instructions)
    TextView instructions;
    @BindView(R.id.spinner_meal)
    Spinner spinner;
    @BindView(R.id.dialog_action_dismiss)
    Button dismiss;
    @BindView(R.id.dialog_action_move)
    Button move;

    private View view;
    private int id;
    private ArrayList<MealDisplayModel> meals;
    private ArrayList<String> mealNames = new ArrayList<>();

    private MealSelectedInDialogListener listener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.dialog_meal_selector, container, false);

        ButterKnife.bind(this, view);

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            id = bundle.getInt("id");
            meals = bundle.getParcelableArrayList("meals");

            for(MealDisplayModel meal: meals)
                mealNames.add(meal.getName());

            if (bundle.getString("instructions") != null) {
                instructions.setText(bundle.getString("instructions"));
                move.setText(getString(R.string.dialog_action_add));
            }
            initializeSpinner();
        }

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                listener = null;
            }
        });

        move.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                listener.mealSelectedInDialog(id, meals.get(spinner.getSelectedItemPosition()));
                listener = null;
            }
        });

        return view;
    }

    private void initializeSpinner() {
        ArrayAdapter<String> mealAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item, mealNames);
        mealAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(mealAdapter);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            listener = (MealSelectedInDialogListener) getParentFragment();
        } catch (ClassCastException e) {
            throw new ClassCastException("fragment must implement MealSelectedInDialogListener");
        }
    }

    public interface MealSelectedInDialogListener {
        void mealSelectedInDialog(int id, MealDisplayModel meal);
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/dialog/MealSelectorDialog.kt
package de.karzek.diettracker.presentation.main.diary.meal.dialog

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatDialogFragment
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.model.MealDisplayModel
import kotlinx.android.synthetic.main.dialog_meal_selector.*
import java.util.*

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
class MealSelectorDialog : AppCompatDialogFragment() {

    private var dialogView: View? = null
    private var currentId: Int = 0
    private var meals: ArrayList<MealDisplayModel>? = null
    private val mealNames = ArrayList<String>()

    private var listener: MealSelectedInDialogListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialogView = inflater.inflate(R.layout.dialog_meal_selector, container, false)

        val bundle = this.arguments
        if (bundle != null) {
            currentId = bundle.getInt("id")
            meals = bundle.getParcelableArrayList("meals")

            for (meal in meals!!)
                mealNames.add(meal.name)

            if (bundle.getString("instructions") != null) {
                instructions.text = bundle.getString("instructions")
                dialog_action_move.text = getString(R.string.dialog_action_add)
            }
            initializeSpinner()
        }

        dialog_action_dismiss.setOnClickListener {
            dismiss()
            listener = null
        }

        dialog_action_move.setOnClickListener {
            dismiss()
            listener!!.mealSelectedInDialog(currentId, meals!![spinner_meal.selectedItemPosition])
            listener = null
        }

        return dialogView
    }

    private fun initializeSpinner() {
        val mealAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, mealNames)
        mealAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner_meal.adapter = mealAdapter
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            listener = parentFragment as MealSelectedInDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("fragment must implement MealSelectedInDialogListener")
        }

    }

    interface MealSelectedInDialogListener {
        fun mealSelectedInDialog(id: Int, meal: MealDisplayModel)
    }
}
